reverse_words/README.rst

```diff
@@ -0,0 +1,119 @@
+Learn to reverse the letters of each individual word in a text
+================================================================
+
+In this demo, a recurrent network equipped with an attention mechanism
+learns to reverse each word (on a character-by-character basis) in its input text. 
+
+The default training data is the Google Billion Word corpus, 
+which you should download and put to the path indicated in your .fuelrc file.
+
+* http://blocks.readthedocs.org/en/latest/
+* http://fuel.readthedocs.org/en/latest/ is practically empty, better to look 
+  at the source : https://github.com/mila-udem/fuel
+
+As an example, the first sentence should be transformed into the second:
+
+* ``The quick brown fox jumps over the lazy dog.`` 
+* ... model ...
+* ``eht kciuq nworb xof spmuj revo eht yzal god.`` 
+
+
+The bulk of the functionality of the code is in the ``__init__.py`` file.
+
+Overall, the model described seems to be as in the paper by Dzmitry Bahdanau, Kyunghyun Cho, Yoshua Bengio (2014) :
+
+.. figure:: 1409.0473v6.Figure1.438x549.png
+   :width: 219px
+   :align: center
+   :alt: Figure 1
+
+   Figure 1 of `Neural Machine Translation by Jointly Learning to Align and Translate <http://arxiv.org/abs/1409.0473>`_
+
+
+Structure of the Data
+--------------------------
+
+The input data arrives through the a ``fuel`` data processing pipeline,
+and the first section of the code deals with functions that clean and prepare the
+data :
+
+* converting character codes to and from a numerical (integer) encoding 
+* a 'gold standard' reverse_words function that performs the task perfectly
+* ``_lower`` and ``_filter_long`` data-cleaning functions that ensure that the 
+  pipeline produces one lower-case character array for each input sentence,
+  and that sentences of over 100 characters are filtered out
+* ``_transpose`` is a convenience function to re-orient the matrices finally coming out of the data stream
+* ``_is_nan`` is a termination function for training - not data processing
```

Comment:
Reading issue #1 , I guess this is a contentious point : How do you want to invoke reverse_words to run?  At the moment ``main.py`` has a broken ``import examples.reverse_words``, so something needs fixing.

Personally, I like the idea of having a ``XYZ/__init__.py`` and ``XYZ/__main__.py`` so that the module can be run from the main repo directory with ``python -m XYZ``.  However, I'm not sure whether I can tell whether the pickling issue would arise there too - partly because there could well be magic under the hood of blocks that I don't know about...

Are you saying that ``_is_nan`` needs to be a function defined at the same 'level' as the class definition of ``WordReverser`` for the model to be pickleable mid-run?  Does this rule out the ``__main__.py`` style invocation in (say) ``markov_chain``?

It would be nice to pick one way to do this, and make it uniform...